lib/operations/command.js

```diff
@@ -44,39 +54,37 @@ class CommandOperation extends OperationBase {
     }
   }
 
-  execute(callback) {
+  execute(command, callback) {
+    if (callback == null) {
+      callback = command;
+      command = this.command;
+    }
     const db = this.db;
-    const command = this.command;
-    const options = this.options;
-    const collection = this.collection || undefined;
+    const options = Object.assign({}, this.options);
 
     // Did the user destroy the topology
-    if (db.serverConfig && db.serverConfig.isDestroyed())
+    if (db.serverConfig && db.serverConfig.isDestroyed()) {
       return callback(new MongoError('topology was destroyed'));
+    }
+
     // Get the db name we are executing against
     const dbName = options.dbName || options.authdb || db.databaseName;
 
     // Convert the readPreference if its not a write
-    if (!this.hasAspect(Aspect.WRITE_OPERATION)) {
-      if (collection != null) {
-        options.readPreference = resolveReadPreference(collection, options);
-      } else {
-        options.readPreference = resolveReadPreference(db, options);
-      }
-    } else {
-      options.readPreference = ReadPreference.primary;
+    if (this.hasAspect(Aspect.WRITE_OPERATION)) {
       applyWriteConcern(command, { db }, options);
```

Comment:
We should be computing `writeConcern` in the constructor, and then just decorating it on to the command here.

Reply:
I could move this `applyWriteConcern` to `_buildCommand`, but `applyWriteConcern` needs the constructed command on which to apply the write concern so it cannot move to the constructor.

Reply:
AFAIK, the only reason the target has to be passed in is for the specific case where the user provides a writeConcern during a transaction. That is normally applicable when the target is an options object, b/c we do not normally let users set specific keys on our commands. The only place I found in our codebase where that `delete target.writeConcern` can potentially be triggered is `mapReduce`.

If we do not need to delete the writeConcern off of the command, then we can run applyWriteConcern on the `options` object, and attach it later during `_buildCommand`:

```js
class CommandOperation extends OperationBase {
  constructor(db, command, options, collection) {
    super(options);
    ...
    this.options = applyWriteConcern(this.options, {db, collection}, this.options);
    ...
  }

  execute(callback) {
    ...
    if (this.options.writeConcern && !this.options.session.inTransaction()) {
      command.writeConcern = this.options.writeConcern;
    }
    ...
  }
}
```

Reply:
`delete target.writeConcern` could also be triggered for a raw command, right? It also looks like it could be triggered for `createCollection`.

Reply:
Transaction spec classifies a `runCommand` helper as a read command https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#generic-runcommand-helper-within-a-transaction. So whether or not we need to do stuff for writeConcern is somewhat ambiguous.

As for `mapReduce` and `createCollection`, we could fix the issue by adding `writeConcern`, `w`, `wtimeout`, and `j` to the exclusion list.